src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs

```diff
@@ -5382,6 +5382,52 @@ public static Task Run(Action action, CancellationToken cancellationToken)
                 TaskCreationOptions.DenyChildAttach, InternalTaskOptions.None);
         }
 
+        public static Task RunAsGreenThread(Action action, CancellationToken cancellationToken)
+        {
+            Task greenThreadTask = new Task(action, cancellationToken);
+            new Thread(()=>
+            {
+                Thread.t_IsGreenThread = true;
+                // TODO? Do something like register the cancellation token as an AsyncLocal or something.
+                greenThreadTask.RunSynchronously();
+            }).Start();
+            return greenThreadTask;
```

Comment:
As with below, this can be something like:
```suggestion
            if (cancellationToken.IsCancellationRequested)
                return Task.FromCanceled(cancellationToken);
                
            var tcs = new TaskCompletionSource();
            new Thread(()=>
            {
                if (cancellationToken.IsCancellationRequested)
                {
                    tcs.SetCanceled(cancellationToken);
                }
                
                Thread.t_IsGreenThread = true;
                try
                {
                    action();
                    tcs.SetResult();
                }
                catch (Exception e)
                {
                    tcs.SetException(e);
                }
            }) { IsBackground = true }.Start();
            return tcs.Task;
```